spec/project/select-modules-in-sources.spec.ts

```diff
@@ -0,0 +1,189 @@
+import { Project, ProjectSource } from '../../core-exports';
+import { ModuleSelectionOptions } from '../../src/project/select-modules-in-sources';
+import { expectToBeValid } from '../model/implementation/validation-utils';
+
+describe('selectModulesInProjectSource', () => {
+    describe('with removeModuleDeclarations = false (default)', () => {
+        it('removes types and fields that are not selected', () => {
+            const result = run(
+                gql`
+                    type OneAndTwo @rootEntity @modules(in: ["module1", "module2"]) {
+                        all: String @modules(all: true)
+                        two: String @modules(in: "module2")
+                        extra1: String @modules(in: ["extra1", "module2 && extra2"])
+                        extra2: String @modules(in: ["extra2"])
+                    }
+
+                    type Two @rootEntity @modules(in: ["module2", "module3"]) {
+                        all: String @modules(all: true)
+                        one: String @modules(in: ["module1"])
+                    }
+
+                    type Three @rootEntity @modules(in: ["module3"]) {
+                        all: String @modules(all: true)
+                        one: String @modules(in: ["module1"])
+                    }
+                `,
+                ['module1', 'module2', 'extra1'],
+            );
+            expect(result).to.equal(`
+                    type OneAndTwo @rootEntity @modules(in: ["module1", "module2"]) {
+                        all: String @modules(all: true)
+                        two: String @modules(in: ["module2"])
+                        extra1: String @modules(in: ["extra1"])
+                        
+                    }
+
+                    type Two @rootEntity @modules(in: ["module2"]) {
+                        all: String @modules(all: true)
+                        one: String @modules(in: ["module1"])
+                    }
+
+                    
+                `);
+        });
+
+        it('removes the non-selected modules from the modules.json', () => {
+            const project = new Project({
+                sources: [
+                    new ProjectSource(
+                        'modules.json',
+                        JSON.stringify({
+                            modules: ['module1', 'module2', 'module3', 'extra1', 'extra2'],
+                            permissionProfiles: [],
+                        }),
+                    ),
+                ],
+                modelOptions: { withModuleDefinitions: true },
+            });
+
+            const result = project.withModuleSelection(['module1', 'module2']);
+            expectToBeValid(result);
+            expect(result.sources.length).to.equal(1);
+            expect(result.sources[0].body).to.equal(`{
+  "modules": [
+    "module1",
+    "module2"
+  ],
+  "permissionProfiles": []
+}`);
+            expect(result.getModel().modules).to.deep.equal([]);
```

Comment:
good catch, that's a bug

Reply:
the project created in the test had an invalid modules.json (it specified permissionProfiles as [] instead of {}), so `.getModel()` silently discarded this file.

Don't want to change `.getModel()` in a breaking way (there could be valid cases where a broken schema is better than none, e.g. during dev), so I now added validation checks to the spec.